### PR TITLE
State Changing events + locking of URLs

### PR DESCRIPTION
**Pull request by mariobyn**

### CODE REVIEW

File: packages/hardhat/contracts/Ethernauts.sol [hunk 1]
```diff
@@ -162,36 +175,47 @@ contract Ethernauts is ERC721Enumerable, Ownable {
...
 
-    function setEarlyMintPrice(uint newEarlyMintPrice) external onlyOwner {
+    function setEarlyMintPrice(uint256 newEarlyMintPrice) external onlyOwner {
         earlyMintPrice = newEarlyMintPrice;
+        emit EarlyMintPriceChanged(newEarlyMintPrice);
     }
 
-    function setBaseURI(string memory newBaseTokenURI) external onlyOwner {
+    function setBaseURI(string calldata newBaseTokenURI) external onlyOwner {
+        require(!permanentUrl, "NFTs minting finished, changing URL not possible");
```
Comment (mjlescano) [inline packages/hardhat/contracts/Ethernauts.sol:189]: This could be handled using [Solidity Errors](https://docs.soliditylang.org/en/v0.8.10/contracts.html?highlight=errors#errors-and-the-revert-statement), instead of requires.

Comment (mariobyn) [reply packages/hardhat/contracts/Ethernauts.sol:189]: Correct, but i said to keep a bit of a pattern? I think the focus is right now to get it out fast asap😅

File: packages/hardhat/hardhat.config.js [hunk 2]
```diff
@@ -8,10 +8,12 @@ require('./tasks/sale-state');
 require('./tasks/mint');
 
 module.exports = {
-  solidity: '0.8.4',
+  solidity: '0.8.9',
   defaultNetwork: 'hardhat',
   networks: {
-    hardhat: {},
+    hardhat: {
+      allowUnlimitedContractSize: true,
```
Comment (mjlescano) [inline packages/hardhat/hardhat.config.js:15]: Why is this necessary?

Comment (mariobyn) [reply packages/hardhat/hardhat.config.js:15]: It isn't

Comment (mariobyn) [reply packages/hardhat/hardhat.config.js:15]: I will remove

File: packages/hardhat/test/Events.test.js [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+
+  before('identify signers', async () => {
+    users = await ethers.getSigners();
+    [owner, user] = users;
+  });
+
+  before('deploy contract', async () => {
+    const factory = await ethers.getContractFactory('Ethernauts');
+
+    const params = Object.assign({}, hre.config.defaults);
```
Comment (mjlescano) [inline packages/hardhat/test/Events.test.js:22]:
```suggestion
    const params = { ...hre.config.defaults };
```

File: packages/hardhat/contracts/Ethernauts.sol [hunk 4]
```diff
@@ -13,35 +13,48 @@ contract Ethernauts is ERC721Enumerable, Ownable {
...
     }
+
     SaleState public currentSaleState;
+    event SaleStateChanged(SaleState state);
+    event BaseTokenURIChanged(string baseTokenURI);
+    event EarlyMintPriceChanged(uint256 earlyMintPrice);
+    event MintPriceChanged(uint256 mintPrice);
+    event CouponSignerChanged(address couponSigner);
+    event WithdrawTriggered(address beneficiary);
+    event PermanentURITriggered(bool value);
```
Comment (slorenzo) [inline packages/hardhat/contracts/Ethernauts.sol:50]: Is this the correct event name?

OpenSea uses `event PermanentURI(string _value, uint256 indexed _id);`
Ref: https://docs.opensea.io/docs/metadata-standards#freezing-metadata

Comment (mariobyn) [reply packages/hardhat/contracts/Ethernauts.sol:50]: Yes, but opensea architecture of their 1155 storefront tokens is diff than ours

Comment (mariobyn) [reply packages/hardhat/contracts/Ethernauts.sol:50]: so this does not apply to us. We need this event just to know when we will lock the changing of URLs once the minting is finished. Once it will be finished, we will be triggering this manually.

File: packages/hardhat/test/Events.test.js [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+      assert.equal(event.args.couponSigner, user.address);
+    });
+  });
+
+  describe('When base URI changes', () => {
+    it('shows BaseTokenURIChanged event is emitted', async () => {
+      receipt = await (await Ethernauts.connect(owner).setBaseURI('http://pinedead.io/')).wait();
+      const event = receipt.events.find((e) => e.event === 'BaseTokenURIChanged');
+      assert.equal(event.args.baseTokenURI, 'http://pinedead.io/');
+    });
```
Comment (slorenzo) [inline packages/hardhat/test/Events.test.js:81]: Nice to have: a test to wait for an error when the owner tries to change the baseURL but the metadata has already frozen.

File: packages/hardhat/test/Events.test.js [hunk 6]
```diff
@@ -0,0 +1,99 @@
...
+
+  describe('When coupon signer changes', () => {
+    it('shows CouponSignerChanged event is emitted', async () => {
+      receipt = await (await Ethernauts.connect(owner).setCouponSigner(user.address)).wait();
+      const event = receipt.events.find((e) => e.event === 'CouponSignerChanged');
+      assert.equal(event.args.couponSigner, user.address);
+    });
+  });
+
+  describe('When base URI changes', () => {
```
Comment (slorenzo) [inline packages/hardhat/test/Events.test.js:76]: Nice to have: A test to expect for an error when the owner tries to change the baseURL but the metadata has already frozen.

Comment (m20cr) [reply packages/hardhat/test/Events.test.js:76]: i will add it as a issue

Comment (mariobyn) [reply packages/hardhat/test/Events.test.js:76]: i will add it as a issue

**Review by mjlescano — APPROVED** [review]
